Add reader for SPSS (.sav) files

cbrnr opened the pull request:
- [x] closes #5768 (at least the reading part, this PR does not cover writing SPSS files)
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

I haven't added a test yet because I wanted to ask which test .sav file I should use (and where to put it). Also, there's no whatsnew entry yet - in which section should this entry go (and I assume it will be out for 0.25.0, so I'll have to change 0.24.3 to 0.25.0 in the docstring).

This PR adds the capability to load SPSS .sav files with `df = pd.io.read_spss("spss_file.sav")`. Currently, there are two optional arguments: `usecols` (should be self-explanatory, let me know if you don't want me to handle a simple `str`) and `categorical`, which maps to the `apply_value_formats` parameter in `read_sav`. With `categorical=True`, a categorical columns is created with the labels from the .sav file. If `False`, numbers will be used.

A few open questions:
- Which additional optional arguments should be made available? [Pyreadstat](https://github.com/Roche/pyreadstat/blob/master/pyreadstat/pyreadstat.pyx#L208-L244) has `dates_as_pandas_datetime`, `encoding`, and `user_missing` which I haven't mapped yet.
- Should the function be called `read_spss` or `read_sav`? SPSS files have the extension `sav`, but the R haven package has a function `read_spss` (which is why I'd prefer `read_spss`).
- Are there any additional meta information bits that could be used/integrated into the data frame? `pyreadstat.read_sav` returns a dataframe and meta-information separately, which I think we shouldn't do in pandas.

cbrnr committed:
Initial version of SPSS reader

cbrnr committed:
Rename file

cbrnr committed:
Add usecols and categorical optional parameters

cbrnr committed:
Fix typo

jreback reviewed (changes requested):
pls add this to the CI in several (but not all places); this should gracefully skip tests if not installed. you can commit the .sav files pandas/tests/io/data/ (see how we do this for .dta). also update the install.rst

cbrnr committed:
Add tests

cbrnr committed:
Skip tests if pyreadstat is not available

cbrnr committed:
Add pyreadstat to Travis (just 37 for now to see if it works)

cbrnr commented:
I've added some test files from the [haven](https://github.com/tidyverse/haven/tree/master/tests/testthat) package - how do we attribute the original authors? They don't have a standard license, so maybe we have to ask them for permission to use their test files?

One test file containing dates cannot be loaded because `pyreadstat` produces an error - this is reported in https://github.com/Roche/pyreadstat/issues/26.

I've updated `install.rst` and I'm currently trying to see if a Travis job successfully completes when `pyreadstat` is installed.

cbrnr committed:
Ignore flake8 F401

cbrnr committed:
Update whatsnew

cbrnr committed:
Change versionadded to 0.25.0

cbrnr committed:
Specify reason for skipif

cbrnr committed:
Fix API

cbrnr committed:
Fix path to test files

cbrnr committed:
Sort imports

TomAugspurger commented:
> I've added some test files from the haven package - how do we attribute the original authors? They don't have a standard license, so maybe we have to ask them for permission to use their test files?

Seems like haven has an MIT license, according to https://cran.r-project.org/web/packages/haven/index.html. Does that sound right @hadley?

If so, I think you can include include Haven's license file in our licenses file, and we'll be good. It'd be good to note the source of these tests files in `test_spss.py`.

hadley commented:
Yeah, that's fine with me. (I generally follow the US and consider data to be un-copyrightable, although in this case I guess it's the specific form that it's important not so much the data)

cbrnr committed:
Use datapath fixture

cbrnr committed:
Acknowledge Haven project

cbrnr committed:
Fix imports order

cbrnr commented:
All tests pass now (currently re-running because of a wrong imports order). Two open questions:

1. I have currently added `pyreadstat` only to one Travis job (3.7). Where would you like this dependency to be added?
2. The Haven license consists of two files ([MIT](https://cran.r-project.org/web/licenses/MIT) and [LICENSE](https://cran.r-project.org/web/packages/haven/LICENSE)). How and where should these files be added (e.g. how should we rename these files)?

jreback commented on ci/deps/travis-37.yaml:
```diff
@@ -22,3 +22,4 @@ dependencies:
   - pip
   - pip:
     - moto
+    - pyreadstat
```
is this only a wheel?

jreback replied:
add to one of the windows builds & the macosx build; is this support on 3.5? any other requirements?

cbrnr replied:
Done.

cbrnr replied:
I don't know what you mean by "only a wheel" though. `pyreadstat` has binary wheels for Windows, Linux, and macOS for Python 3.5, 3.6, and 3.7: https://pypi.org/project/pyreadstat/#files

jreback replied:
meaning you are installing thru pip, rather use a conda package if its available

jreback commented on doc/source/whatsnew/v0.25.0.rst:
```diff
@@ -265,6 +265,8 @@ Optional libraries below the lowest tested version may still work, but are not c
 +-----------------+-----------------+
 | pyarrow         | 0.9.0           |
 +-----------------+-----------------+
+| pyreadstat      | 0.2.5           |
```
this is not needed here; this is only for changes

cbrnr replied:
Done.

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,27 @@
+def read_spss(path, usecols=None, categorical=True):
```
can you add typing

cbrnr replied:
Sure. This is the first time I've used typing so please double-check.

jreback commented on pandas/tests/io/test_spss.py:
```diff
@@ -0,0 +1,77 @@
+from pandas.io.spss import read_spss
+
+try:
+    import pyreadstat  # noqa
```
use ``pyreadstat = pytest.importorskip('pyreadstat')``; no need for the skipifs

cbrnr replied:
Done.

jreback reviewed (changes requested):
put licenses in pandas/LICENSES

cbrnr committed:
Use importorskip

cbrnr committed:
Add pyreadstat dependency to macOS and Windows CI

cbrnr committed:
Add typing

pep8speaks commented:
Hello @cbrnr! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:





* In the file [`pandas/io/spss.py`](https://github.com/pandas-dev/pandas/blob/3305d370b0ce84f35b47adce3e9e8bd375938f83/pandas/io/spss.py):

> [Line 6:55](https://github.com/pandas-dev/pandas/blob/3305d370b0ce84f35b47adce3e9e8bd375938f83/pandas/io/spss.py#L6): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 6:56](https://github.com/pandas-dev/pandas/blob/3305d370b0ce84f35b47adce3e9e8bd375938f83/pandas/io/spss.py#L6): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 7:32](https://github.com/pandas-dev/pandas/blob/3305d370b0ce84f35b47adce3e9e8bd375938f83/pandas/io/spss.py#L7): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals
> [Line 7:33](https://github.com/pandas-dev/pandas/blob/3305d370b0ce84f35b47adce3e9e8bd375938f83/pandas/io/spss.py#L7): [E252](https://duckduckgo.com/?q=pep8%20E252) missing whitespace around parameter equals

cbrnr committed:
Add missing whitespace

cbrnr committed:
Add Haven license files

cbrnr commented:
@jreback I've implemented all changes.

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def read_spss(path: str,
+              usecols: Union[str, Sequence[str], None] = None,
```
this can be a string?

we need to define a ListLike in pandas._typing I think (doesn't have to be done here)

cbrnr replied:
Standard `typing` has `Sequence` and `Iterator` (and `List` for return types), but I agree that a duck type `ListLike` would be useful.

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,33 @@
+
+    Parameters
+    ----------
+    path : string
```
does this accept a pathlike? Union[str, Path] ?

cbrnr replied:
`pathlib.Path` should work everywhere a path string is expected, so IMO it is not really necessary to explicitly add this. But if you want I can of course add it (this requires an extra `import pathlib` though).

cbrnr replied:
I've implemented this change.

cbrnr commented:
Yes, `pyreadstat` is only available via `pip` and not via `conda`.

cbrnr committed:
Remove trailing whitespace

cbrnr committed:
Fix import format and add pathlib.Path

cbrnr commented:
Fixed the formatting issues, hopefully this will come back green and then it should be ready to merge.

WillAyd reviewed (changes requested):
Minor nit on annotation but otherwise this lgtm - nice change!

cbrnr commented:
@WillAyd could you please elaborate?

WillAyd commented:
Not sure what happened to my comment but was asking to change `Union[str, Sequence[str], None]` to `Optional[Union[str, Sequence[str]]`

cbrnr committed:
Use Optional to properly type an optional argument with default value of None

cbrnr commented:
@WillAyd done!

TomAugspurger commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,35 @@
+    DataFrame
+    """
+
+    from pyreadstat import read_sav
```
Typically we do these in a `try / except ImportError` so we can raise with a nicer message that pyreadstat is an optional dependency.

cbrnr replied:
Done.

TomAugspurger commented on pandas/tests/io/test_spss.py:
```diff
@@ -0,0 +1,68 @@
+    # test file from the Haven project (https://haven.tidyverse.org/)
+    fname = datapath("io", "data", "labelled-num.sav")
+
+    df = read_spss(fname, categorical=True)
```
style nitpick: can you use `pd.read_spss` here? That'll better test the user-facing behavior.

cbrnr replied:
Done.

cbrnr committed:
Use conda-forge instead of PyPI

ofajardo commented:
Pyreadstat is available both with pip and conda. In the README that is explained in the How to install section

cbrnr committed:
Better ImportError message

cbrnr committed:
Use pd.read_spss

cbrnr commented:
I didn't see that you have conda-forge added, sorry about that. I've addressed all comments.

cbrnr committed:
Add conda-forge on macOS task

cbrnr commented:
No idea what's going on with Azure, could someone please restart?

jreback commented:
can you merge master (you have a conflict)

cbrnr commented:
OK, I've rebased.

cbrnr committed:
Use pyreadstat from pip for Python 3.5 (not available on conda-forge)

jreback commented:
looks good. @jorisvandenbossche @TomAugspurger @bashtage ok with this?

bashtage commented:
The only comment I would make is based on helping the Stata reader: are the SAV files comprehensive in the sense that they contain every combination of data storage features? This was an issue with the Stata reader where the original test files did not contain all features which lead to a slow but stead set of error reports, until finally a set was created that had every possible storage format and encoding.

cbrnr commented:
I've taken the test files from the Haven R package, which uses the same underlying C library [readstat](https://github.com/WizardMac/ReadStat) to load SAV files. One of these files contains a strange timestamp (year 201416), which currently produces an error when read with pyreadstat (but readstat as well as Haven can read this file). @ofajardo has already fixed this by adding a new parameter in https://github.com/Roche/pyreadstat/issues/26. Once he makes a new release, I can add a new parameter `disable_datetime_conversion ` to `pd.read_spss` (then we need to require at least this version). Alternatively, I could add `kwargs` which get passed to the underlying pyreadstat function (and we don't check anything on the pandas side).

There's also a second test file I didn't use from the Haven package, because it contains value labels which currently cannot be converted to something pandas offers. The file can be read successfully though.

cbrnr commented:
Regarding the second test file: SAV files can have column names and column labels. pyreadstat can read both, but we have to set either one of the two as `df.columns`. There is a parameter that controls which gets used, but I assume pandas doesn't have a place within the `DataFrame` object to store both column names and column labels?

ofajardo commented:
@cbrnr regarding the second test file, which variable labels cannot be converted to pandas? I think pyreadstat is getting everything that the C Readstat offers, but you can have either the original value or the labelled value but not both (similar situation as with  column labels), this can be handled passing keyword arguments to pyreadstat.

cbrnr commented:
Yes, that's what I meant - you cannot have both since a DataFrame presumably cannot store both entities (or can it?).

cbrnr commented:
@jreback @jorisvandenbossche @TomAugspurger @bashtage do you want me to make any additional changes?

jreback commented on ci/deps/travis-37.yaml:
```diff
@@ -20,5 +20,6 @@ dependencies:
   - s3fs
   - isort
   - pip
+  - pyreadstat
```
can you also add to environment.yml (note there is a PR changing it as we speak, so might have a conflict)

cbrnr replied:
Add it via conda or pip (conda has the package only for Python 3.6 and 3.7, but not for 3.5)?

cbrnr replied:
I've used pip because I guess we want that to work on Python 3.5.

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,37 @@
+
+
+def read_spss(path: Union[str, Path],
+              usecols: Optional[Union[str, Sequence[str]]] = None,
```
I don't think this is correct, usecols (in all of our other parsers) must be a list-like or None.

can you change this (and add a test for non-list-like)

jreback replied:
@bashtage we have the columns arg on read_stata, this is simliar, yes? (maybe should change read_stata to accept usecols rather than columns as its more consistent with the other IO readers)

cbrnr replied:
`pyreadstat` requires `usecols` to be a `list` (`tuple` or other list-likes raise a `TypeError`). Do you want me to just re-raise this exception, or should I try to convert list-likes to a `list` and raise a custom exception if this is not possible?

TomAugspurger replied:
Ideally we would accept any sequence and convert it to a list for pyreadstat.

cbrnr replied:
Done, `usecols = list(usecols)` should give a clear error message if `usecols` is not a sequence.

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,37 @@
+
+def read_spss(path: Union[str, Path],
+              usecols: Optional[Union[str, Sequence[str]]] = None,
+              categorical: bool = True) -> DataFrame:
```
can we can call this ``convert_categoricals`` (similar to read_stata)

cbrnr replied:
Done.

cbrnr committed:
usecols only accepts list-like or None

cbrnr committed:
Remove condition (str is not allowed anymore)

cbrnr committed:
Rename to convert_categoricals

cbrnr committed:
Explicitly convert to list

cbrnr committed:
Add pyreadstat to environment.yml

cbrnr commented:
@TomAugspurger I've addressed your comments.

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,37 @@
+    except ImportError:
+        raise ImportError("pyreadstat is required to read SPSS .sav files.")
+
+    usecols = list(usecols)  # explicitly convert to list
```
actually you need to check

```
if usecols is not None:
  if not is_list_like(usecols):
      ....raise nice message
```

cbrnr replied:
Fixed.

jreback commented on pandas/tests/io/test_spss.py:
```diff
@@ -0,0 +1,78 @@
+    msg = ("Argument 'usecols' has incorrect type "
+           "(expected list, got {})".format(type(usecols).__name__))
+
+    with pytest.raises(TypeError, match=msg):
```
suprised this doesn't fail, as this turns *into* a list (see my above comment)

cbrnr replied:
Fixed.

jreback reviewed (changes requested):
pls merge master as well

cbrnr committed:
Use is_list_like

cbrnr commented:
Rebased and fixed comments, this should be good to go.

TomAugspurger commented:
Some CI failures
https://travis-ci.org/pandas-dev/pandas/jobs/543229555#L2502

On Sat, Jun 8, 2019 at 3:29 PM Clemens Brunner <notifications@github.com>
wrote:

> Rebased and fixed comments, this should be good to go.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/26537?email_source=notifications&email_token=AAKAOIQNM2BDUQ2MYTCKVHLPZQJD3A5CNFSM4HP3GQF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODXH5PJA#issuecomment-500160420>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIX6JEVW3XAIOZXTAR3PZQJD3ANCNFSM4HP3GQFQ>
> .
>

cbrnr committed:
Fix tests

cbrnr commented:
Fixed.

cbrnr committed:
Fix df is assigned but never used

TomAugspurger reviewed (commented):
Small lint failure. LGTM otherwise https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=12586

cbrnr committed:
Sort imports

cbrnr commented:
Hopefully fixed now.

cbrnr committed:
Update requirements-dev.txt

cbrnr commented:
The failure on Windows was hopefully not related (test_sorting.py) because this passes locally. Let's see how it goes.

jreback commented on ci/deps/azure-windows-37.yaml:
```diff
@@ -30,3 +30,5 @@ dependencies:
   - pytest-mock
   - moto
   - hypothesis>=3.58.0
+  - isort
```
can you remove isort

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,42 @@
+    ----------
+    path : string or Path
+        File path
+    usecols : list-like or None
```
list-like, optional

jreback commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,42 @@
+        File path
+    usecols : list-like or None
+        Return a subset of the columns. If None, return all columns.
+    convert_categoricals : bool
```
bool, default is True

jreback reviewed (changes requested):
small comments, ping on green.

cbrnr committed:
Remove isort

cbrnr committed:
Improve docstring

cbrnr committed:
Revert indent

cbrnr committed:
Indent should be 2 spaces

cbrnr commented:
@jreback all green!

jreback commented on doc/source/install.rst:
```diff
@@ -273,6 +273,7 @@ Optional Dependencies
       for writing .xlsx files (xlrd >= 1.0.0)
     * `XlsxWriter <https://pypi.org/project/XlsxWriter>`__: Alternative Excel writer
 
+* `pyreadstat <https://github.com/Roche/pyreadstat>`__: required for reading SPSS .sav files.
```
this changed to a table format

cbrnr replied:
Fixed.

jreback commented on requirements-dev.txt:
```diff
@@ -52,4 +52,5 @@ sqlalchemy
-xlwt
\ No newline at end of file
+xlwt
+pyreadstat
```
can you put in the correct alphabetical place

cbrnr replied:
The file wasn't sorted at all. I've sorted all lines now.

jreback reviewed (changes requested):
install.rst changed, can you merge master and update for the conflict; ping on green.

TomAugspurger commented on pandas/io/spss.py:
```diff
@@ -0,0 +1,42 @@
+    -------
+    DataFrame
+    """
+    try:
```
Apologies for the merge conflict, I just refactored this. You should be able to replace this try / except with `import_optional_dependency` from `pandas.compat._optional`:

http://pandas-docs.github.io/pandas-docs-travis/development/contributing.html#optional-dependencies

Can you pick a minimum version for pyreadstat that we should require? Add that to the `VERSIONS` dict in `pandas.compat._optional`.

cbrnr replied:
Done, except for there is a problem with different version tags in pyreadstat: https://github.com/Roche/pyreadstat/issues/28

Do you want to wait if @ofajardo has time to cut a new release where this is fixed (will be 0.2.6 then), or should we remove the minimum requirement for now?

cbrnr committed:
Add minimum version for pyreadstat

cbrnr committed:
Add minimum version

cbrnr committed:
Use import_optional_dependency

cbrnr committed:
Remove minimum version for now

cbrnr committed:
Correct import order

cbrnr commented:
@jreback @TomAugspurger everything green!

jreback commented:
@cbrnr one more conflict, ping on green.

cbrnr committed:
Remove duplicate

cbrnr committed:
Don't need conda-forge here

cbrnr committed:
Remove blank line

cbrnr committed:
Fix order

cbrnr commented:
Finally all green!

jreback commented:
thanks @cbrnr

ofajardo commented:
@cbrnr I am in the process of releasing a new version of pyreadstat with writing capabilities, in case that's of your interest. Should be there later today or tomorrow.

cbrnr commented:
Nice! However, at the moment I don't think there's a need for exporting to a proprietary format directly from pandas. If someone really wants to do that they can use your package directly.

isantolin commented:
Still missing documentation http://pandas-docs.github.io/pandas-docs-travis/user_guide/io.html

TomAugspurger commented:
Can you open a PR fixing that? Or a new issue if you don't plan to?

On Fri, Jul 19, 2019 at 8:16 AM Ignacio Santolin <notifications@github.com>
wrote:

> Still missing documentation
> http://pandas-docs.github.io/pandas-docs-travis/user_guide/io.html
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pandas-dev/pandas/pull/26537?email_source=notifications&email_token=AAKAOIUY5EKTLKJB46MV5M3QAG5DRA5CNFSM4HP3GQF2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD2LS6KY#issuecomment-513224491>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAKAOIU5ZPA6JGX7HGCVAITQAG5DRANCNFSM4HP3GQFQ>
> .
>
